Sources/_StringProcessing/Regex/AnyRegexOutput.swift

```diff
@@ -232,21 +182,81 @@ extension Regex where Output == AnyRegexOutput {
   public init<Output>(_ regex: Regex<Output>) {
     self.init(node: regex.root)
   }
+}
 
-  /// Returns a typed regex by converting the underlying types.
+@available(SwiftStdlib 5.7, *)
+extension Regex {
+  /// Creates a strongly-typed regex from a dynamic regex, i.e. one created
+  /// from a string.
   ///
-  /// - Parameter type: The expected output type.
-  /// - Returns: A regex generic over the output type if the underlying types can be converted.
-  ///   Returns `nil` otherwise.
-  public func `as`<Output>(
-    _ type: Output.Type = Output.self
-  ) -> Regex<Output>? {
-    let result = Regex<Output>(node: root)
-    
-    guard result._verifyType() else {
+  /// Use this initializer to create a strongly typed regex from
+  /// one that was created from a string. Returns `nil` if the types
+  /// don't match.
+  public init?(_ dynamic: Regex<AnyRegexOutput>) {
```

Comment:
It feels like we should have a `as: Output.Type = Output.self` parameter to match the run-time string-based initializer. e.g.

```swift
let pattern = "a+(b+)c+"
let a = try Regex(pattern, as: (Substring, Substring).self))
let b: Regex<(Substring, Substring)> = try Regex(pattern)

let dynamic = try Regex(pattern)
let c = Regex(dynamic, as: (Substring, Substring).self)!
let d: Regex<(Substring, Substring)> = Regex(dynamic)!
```

Reply:
Argh that syntax highlighting. Any ideas for a better argument name than `dynamic`?